File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;
class Solution {
    public int solution(int k, int[][] dungeons) {
       
        return explore(k, dungeons, new boolean[dungeons.length]);
    }
    
    private int explore(int k, int[][] dungeons, boolean[] visited){
        int maxDungeons = 0;
        
        for(int i = 0 ; i < dungeons.length; i++){
            if(!visited[i] && k >= dungeons[i][0]){
                visited[i] = true;
                int dungeonExplored = explore(k - dungeons[i][1], dungeons, visited) + 1;
                maxDungeons = Math.max(maxDungeons, dungeonExplored);
                visited[i] = false;
            }
        }
        
        return maxDungeons;
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    
    function maxDungeons(k,dungeons,visited){
        let maxDungeonCount = 0;
        for(let i = 0 ; i < dungeons.length ; i++){
            if(dungeons[i][0] <= k && !visited[i]){
                visited[i] = true;
                maxDungeonCount = Math.max(maxDungeonCount ,maxDungeons(k - dungeons[i][1], dungeons , visited) + 1);
                visited[i] = false;
            }
        }

        return maxDungeonCount;
    }
    
    let visited = Array.from({ length: dungeons.length }, ()=>false);
    return maxDungeons(k,dungeons,visited)
}